### PR TITLE
Simplify inline-r threading model

### PR DESCRIPTION
**Pull request by mboes**
This patch series removes a lot of code. Notably, we remove the concept
of R thread. R computations now always take place in the same thread as
the caller. It is up to the caller to ensure single-threaded access, but
thanks to the `R` monad that's easy because `forkIO` is not expressible.

The cost here is that we lose event processing in GHCi sessions for H.
However, this only worked reliably on Linux, and was known to be broken
on other at least OS X. It was a hack, with a lot of funny side effects,
like R being continuously confused thinking it has no stack space left.
So we had to forcibly disable stack checks. This should now be
superfluous.

In GHCi, users should call `H.refresh` whenever they want to repaint
a window, or relinquish control to an event loop.

cc @facundominguez @Fuuzetsu.

### CODE REVIEW

**Comment by Fuuzetsu** [comment]
`Language.R.Event.refresh` should probably by re-exported under `H`

**Comment by qnikst** [comment]
With this PR, R with `--enable-strict-barrier` segfaulting on it's own tests with different memory management related messages:

```
stack test
inline-r-0.6.0.1: build (test)
H-0.6.0.1: test (suite: tests)
Progress: 1/2*** longjmp causes uninitialized stack frame ***: /usr/lib64/ghc-7.8.4/bin/ghc terminated
======= Backtrace: =========
/lib64/libc.so.6(+0x73d6f)[0x7f746791ad6f]
/lib64/libc.so.6(__fortify_fail+0x37)[0x7f746799fb67]
/lib64/libc.so.6(+0xf8a9d)[0x7f746799fa9d]
/lib64/libc.so.6(__longjmp_chk+0x29)[
...
/usr/lib64/R/lib/libR.so(R_ToplevelExec+0xfe)[0x7f7426eeefae]
/usr/lib64/R/lib/libR.so(R_tryEval+0x29)[0x7f7426eef029]
/usr/lib64/R/lib/libR.so(R_tryEvalSilent+0x1a)[0x7f7426eef07a]
```

```
   [2 of 6] Compiling Test.Regions     ( tests/Test/Regions.hs, .stack-work/dist/x86_64-linux/Cabal-1.18.1.5/build/tests/tests-tmp/Test/Regions.o )
    
     *** caught segfault ***
    address 0x8ec0, cause 'memory not mapped'
    aborting ...
    Error: unprotect_ptr: pointer not found
    Execution halted
```

**Comment by qnikst** [comment]
I'm not sure that I can say why we have following errors, it's may be one of:

1. we initialize R constants from non-bounded thread thus constants that we are `poking`
are not correct.

2. we are using R concurrently (this could happen in QQ when we are compiling).

**Comment by Fuuzetsu** [comment]
`Control.Concurrent.OSThread` can be removed

**Comment by qnikst** [comment]
Fails on ghc-7.10 with
```
src/Language/R/Event.hs:87:78:
        Not in scope: data constructor ‘Event.MultiShot’
```

File: inline-r/src/Language/R/Event.hs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+registerREvents emgr = io $ do
+    tmgr <- Event.getSystemTimerManager
+    fdkeys <- forIH inputHandlers $ \R.InputHandler{..} -> do
+      let action _ _ = invokeCallback inputHandlerCallback inputHandlerUserData
+      case 0 < inputHandlerActive of
+        True ->
+#if MIN_VERSION_base(4,8,0)
+          Just <$> Event.registerFd emgr action inputHandlerFD Event.evtRead Event.MultiShot
+#else
+          Just <$> Event.registerFd emgr action inputHandlerFD Event.evtRead
```
Comment (facundominguez) [inline inline-r/src/Language/R/Event.hs]: How does this ensure that the input handlers are run in the main thread?

`registerREvents` doesn't seem to be used or tested anywhere.

Comment (mboes) [inline inline-r/src/Language/R/Event.hs]: It doesn't. As the module docstring says, it just allows you to mesh with an arbitrary event loop that you provide it as an argument. It's up to you to ensure that the given loop is run on the main thread. While in GHC 7.6 it was possible to launch an event manager loop, in 7.8 and 7.10 the definition is not exported (likely an oversight, see email to list). So it's not possible to test on the main thread on these compilers. I guess a test could be constructed that uses a system event manager whose event loop is already running, but that wouldn't be on the main thread (not a problem for certain use cases like using R as a webserver).

Comment (facundominguez) [inline inline-r/src/Language/R/Event.hs]: I see. If this code cannot be executed in latest compilers, how do you know it is good?

File: inline-r/src/Foreign/R/EventLoop.chs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import System.Posix.Types (Fd(..))
+import Prelude -- Silence AMP warning.
+
+#include <R_ext/eventloop.h>
+
+-- | R input handler chain. Each input handler points to the next. This view of
+-- input handlers is /shallow/, in the sense that the 'Storable' instance only
+-- unmarshalls the first element in the chain at any one time. A shallow view
+-- allows 'peek' and 'poke' to be inlinable.
+data InputHandler = InputHandler
```
Comment (facundominguez) [inline inline-r/src/Foreign/R/EventLoop.chs]: Does marshaling input handlers bring any ability to r-inline? I cannot think of a use case.

File: inline-r/src/Data/Vector/SEXP/Mutable.chs [hunk 3]
```diff
@@ -313,8 +313,7 @@ unsafeToStorable :: (PrimMonad m, VECTOR s ty a)
 {-# INLINE unsafeToStorable #-}
 unsafeToStorable v@(MVector p) = unsafePrimToPrim $ do
   R.preserveObject p
-  post <- getPostToCurrentRThread
-  ptr <- newForeignPtr (toVecPtr v) (post $ R.releaseObject (R.sexp $ castPtr $ toVecPtr v))
+  ptr <- newForeignPtr (toVecPtr v) (R.releaseObject (R.sexp $ castPtr $ toVecPtr v))
```
Comment (facundominguez) [inline inline-r/src/Data/Vector/SEXP/Mutable.chs]: Is it safe to have the GC call R functions in the finalizer?

**Comment by facundominguez** [comment]
The user's guide should be updated to explain the restriction of running R functions or using `[r|...]` in the main thread, and it will need a section on how to deal with graphics as event loops were not needed before.

Running in other threads than the main thread was possible with some hacks in the past. If using other threads is declared legal, these hacks should be documented unless they are no longer necessary. Which brings another issue: Tests should be run both in the main thread and in other threads when it is supposed to work.

File: inline-r/src/Foreign/R/EventLoop.chs (hunk 2 above)
Comment (mboes) [inline inline-r/src/Foreign/R/EventLoop.chs]: Yes - it makes it possible to implement e.g. `registerREvents`, or any other function that wants to enumerate R input handlers to incorporate these handlers into some other event loop iteration function. The alternative is a bunch of accessor functions, but why use that when you can create a cheap view of an R data structure which most of the time gets inlined and simplified away anyways.

**Comment by mboes** [comment]
Updates to documentation on gh-pages branch pending. Merging since this is a blocker for release.

**Comment by Fuuzetsu** [comment]
Is the documentation now in release-ready state by the way?

**Comment by mboes** [comment]
Needs work still. A few out-of-date sections.

**Pull request merged**